rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,105 @@
+
+### Teleport Auth server discovery and authentication
+
+To write something into Teleport, the manager needs to know what server to connect to and how. There're many options on how to discover Teleport instances in the cluster but for now, let's start with the simplest one - run manager as a sidecar container in the same pod as the Teleport auth server. Containers in the pod can share the volumes, so the manager container could have access to `/var/lib/teleport/proc` database, which contains administrator credentials. This is the same way how `tctl` utility works - it doesn't require any authentication when you run it on the same machine where Teleport is running.
+
+### Resource definitions
+
+Custom resource definitions of Teleport resource look like this:
+
+```yaml
+apiVersion: resources.goteleport.com/v2
+kind: Role
+metadata:
+  name: folks
+  namespace: teleport
+instanceName: some-instance
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+instanceName: some-instance
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport. Every custom resource must have the `instanceName` field useful to look up an instance resource in order to know what address to connect to. Instance resource and its associated secret must reside in the same namespace as the custom resource.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields, `instanceName` in particular. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### Identity definitions
+
+To run some plugin in the cluster, the credentials are required. Normally, one can get the credentials credentials using `tctl auth sign` command or using `tsh login`. To automate the credentials generation, the `Identity` resource is introduced. The identity definition looks like this:
+
+```yaml
+apiVersion: control.goteleport.com/v1
+kind: Identity
+metadata:
+  name: access-slack
+spec:
+  username: access-slack
+  secretName: access-slack-secret
+```
```

Comment:
How is the Identity resource used by the operator? If it's running as a side-car, it should be able to talk to auth server directly?

Reply:
It's useful to generate the credentials for plugins.
Plugins themselves could run as a sidecar but I don't find it a good idea for security reasons.